### PR TITLE
WA-186: Set device password and create account

### PR DESCRIPTION
**Pull request by fmrsabino**
* New designs for password and fingerprint setup
* Removed account creation related screens (account creation is now done in the background with no user interaction)
* Update Push Notification Payloads 

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/res/layout/layout_onboarding_intro.xml [hunk 1]
```diff
@@ -1,39 +1,77 @@
...
+            app:layout_constraintTop_toTopOf="parent" />
 
         <TextView
-            android:layout_width="match_parent"
+            android:id="@+id/textView"
+            android:layout_width="wrap_content"
             android:layout_height="wrap_content"
-            android:layout_marginTop="24dp"
+            android:layout_marginBottom="24dp"
+            android:layout_marginEnd="70dp"
```
Comment (rmeissner) [inline app/src/main/res/layout/layout_onboarding_intro.xml]: these margins should be different if we follow the designs

File: app/src/main/res/layout/layout_onboarding_intro.xml [hunk 2]
```diff
@@ -1,39 +1,77 @@
...
-        android:orientation="vertical"
-        android:padding="16dp">
+        android:layout_height="match_parent"
+        android:background="@drawable/onbarding_intro_gradient"
+        android:gravity="center_horizontal">
 
         <ImageView
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:background="@drawable/ic_onboarding_intro_background"
```
Comment (rmeissner) [inline app/src/main/res/layout/layout_onboarding_intro.xml]: would really prefer if this is a svg, as this is a really big image

If too complicated this is fine

File: app/src/main/res/layout/layout_onboarding_intro.xml [hunk 3]
```diff
@@ -1,39 +1,77 @@
...
+            android:scaleType="centerCrop"
```
Comment (rmeissner) [inline app/src/main/res/layout/layout_onboarding_intro.xml]: center crop might not be enough for really high res phones, right? (SVG would solve this :P )

File: app/src/main/res/layout/layout_password_confirm.xml [hunk 4]
```diff
@@ -18,67 +19,105 @@
...
                 </android.support.design.widget.TextInputLayout>
 
+                <TextView
+                    android:id="@+id/layout_password_confirm_info"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:paddingEnd="16dp"
+                    android:paddingStart="16dp"
+                    android:text="@string/password_doesnt_match"
+                    android:textColor="#f02525" />
```
Comment (rmeissner) [inline app/src/main/res/layout/layout_password_confirm.xml]: please add to colors

File: app/src/main/res/values-v21/styles.xml [hunk 5]
```diff
@@ -9,4 +9,10 @@
         <item name="android:statusBarColor">@color/gnosis_blue</item>
         <item name="android:navigationBarColor">@color/gnosis_blue</item>
     </style>
+
+    <style name="IntroTheme" parent="IntroThemeBase">
+        <item name="android:statusBarColor">#00b8e0</item>
```
Comment (rmeissner) [inline app/src/main/res/values-v21/styles.xml]: can we have this in the color.xml

File: app/src/main/res/values/colors.xml [hunk 6]
```diff
@@ -54,4 +54,16 @@
     <color name="security_bar_low">#eab41c</color>
     <color name="security_bar_good">#09a862</color>
     <color name="security_bar_best">#10b3d4</color>
+
+    <!-- New Design colors -->
+    <color name="green_new">#09a862</color>
```
Comment (rmeissner) [inline app/src/main/res/values/colors.xml]: Can we use the names from zeplin pls, makes it easier to figure out which color it is

File: app/src/main/res/values/styles.xml [hunk 7]
```diff
@@ -71,6 +71,51 @@
...
+        <item name="android:paddingTop">8dp</item>
+        <item name="android:paddingBottom">8dp</item>
+        <item name="android:gravity">center</item>
+    </style>
+
+    <!-- White theme (will be the main theme for the app) -->
+
+    <style name="WhiteThemeBase" parent="AppTheme">
+        <item name="android:windowBackground">@color/white</item>
+        <item name="colorAccent">#28b2fa</item>
```
Comment (rmeissner) [inline app/src/main/res/values/styles.xml]: pls add to the colors.xml

File: app/src/main/res/values/styles.xml [hunk 8]
```diff
@@ -71,6 +71,51 @@
...
+    <!-- Intro theme. -->
+
+    <style name="IntroThemeBase" parent="AppTheme">
+        <item name="android:windowBackground">@drawable/onbarding_intro_gradient</item>
+    </style>
+
+    <style name="IntroTheme" parent="IntroThemeBase" />
+
+    <style name="SmallButton" parent="Theme.AppCompat.Light">
+        <item name="colorButtonNormal">#28b2fa</item>
```
Comment (rmeissner) [inline app/src/main/res/values/styles.xml]: pls add these all to the colors.xml

File: app/src/main/java/pm/gnosis/heimdall/ui/onboarding/OnboardingIntro.kt [hunk 9]
```diff
@@ -16,19 +18,32 @@ import timber.log.Timber
...
+        bottomSheetBehavior = BottomSheetBehavior.from(layout_onboarding_intro_bottom_sheet).apply { state = BottomSheetBehavior.STATE_HIDDEN }
     }
 
     override fun onStart() {
         super.onStart()
         disposables += layout_onboarding_intro_get_started.clicks()
             .subscribeBy(
-                onNext = { startActivity(PasswordSetupActivity.createIntent(this)) },
-                onError = Timber::e
+                onNext = {
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply app/src/main/java/pm/gnosis/heimdall/ui/onboarding/OnboardingIntro.kt]: No specific reason

File: app/src/main/java/pm/gnosis/heimdall/utils/ViewUtils.kt [hunk 10]
```diff
@@ -21,3 +22,13 @@ fun TextView.setFormattedText(res: Int, vararg params: Pair<String, String>) {
 }
 
 fun ImageView.setColorFilterCompat(@ColorRes color: Int) = setColorFilter(context.getColorCompat(color))
+// TODO: remove method after extracting colors
+fun ImageView.setColorFilterCompat2(color: Int) = setColorFilter(color)
+
+fun TextView.setSelectedCompoundDrawablesWithIntrinsicBounds(
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply app/src/main/java/pm/gnosis/heimdall/utils/ViewUtils.kt]: Was thinking of a name for a function with optional parameters. Suggestions? 😛

File: app/src/main/java/pm/gnosis/heimdall/ui/onboarding/password/PasswordConfirmActivity.kt [hunk 11]
```diff
@@ -47,40 +51,74 @@ class PasswordConfirmActivity : SecuredBaseActivity() {
...
+        )
     }
 
     private fun onPasswordSetupError(throwable: Throwable) {
         Timber.e(throwable)
         (throwable as? PasswordInvalidException)?.let { passwordInvalidException ->
             when (passwordInvalidException.reason) {
-                is PasswordNotLongEnough -> layout_password_confirm_input_layout.error = getString(R.string.password_too_short)
-                is PasswordsNotEqual -> layout_password_confirm_input_layout.error = getString(R.string.passwords_do_not_match)
+                is PasswordNotLongEnough -> snackbar(layout_password_confirm_coordinator, R.string.password_too_short)
```
[the remark this answers is not in the document]
Comment (fmrsabino) [reply app/src/main/java/pm/gnosis/heimdall/ui/onboarding/password/PasswordConfirmActivity.kt]: This should never happen no. Either we "trust" that the previous activity to correctly validate the input or we repeat the checks in the confirmation phase (not doing it either).

File: app/src/main/res/layout/layout_onboarding_intro.xml (hunk 2 above)
Comment (fmrsabino) [reply app/src/main/res/layout/layout_onboarding_intro.xml]: For now SVG due to the problem with the current one.

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -384,4 +384,8 @@
     <string name="password_validation_identical_characters">No more than 2 identical characters in a row\n</string>
     <string name="password_validation_minimum_characters">Use a minimum of 8 characters\n</string>
     <string name="password_validation_one_number_one_letter">Password with at least 1 number and 1 letter\n</string>
+    <string name="terms">Terms</string>
+    <string name="terms_info"><![CDATA[By continuing you accept and agree with the <a href="https://gnosis.pm">Terms of Use</a> and <a href="https://gnosis.pm/resources/default/pdf/PrivacyPolicyGnosisLtd.pdf">Privacy Policy</a>]]></string>
+    <string name="no_thanks">NO THANKS</string>
```
Comment (rmeissner) [inline app/src/main/res/values/strings.xml]: I normally use all caps in the textview, so I can reuse it if necessary ... but doesn't really matter :D

File: app/src/main/res/values/styles.xml [hunk 13]
```diff
@@ -71,6 +71,62 @@
...
+
+    <style name="SmallButton" parent="Theme.AppCompat.Light">
+        <item name="colorButtonNormal">@color/azure</item>
+        <item name="android:textColor">@color/white</item>
+        <item name="android:paddingStart">40dp</item>
+        <item name="android:paddingEnd">40dp</item>
+    </style>
+
+    <style name="SmallFlatButton" parent="Theme.AppCompat.Light">
+        <item name="colorControlHighlight">#28b2fa</item>
```
Comment (rmeissner) [inline app/src/main/res/values/styles.xml]: colors.xml

**Comment by rmeissner** [comment]
Maybe cleanup the AndroidManifest. It has some non-existing activities, right?

File: app/src/main/java/pm/gnosis/heimdall/reporting/EventTracker.kt [hunk 14]
```diff
@@ -31,6 +31,7 @@ enum class ScreenId {
     CREATE_TRANSACTION,
     CHANGE_PASSWORD,
     DEBUG_SETTINGS,
```
Comment (rmeissner) [inline app/src/main/java/pm/gnosis/heimdall/reporting/EventTracker.kt]: We should clean up these identifiers

File: app/src/main/java/pm/gnosis/heimdall/ui/onboarding/password/PasswordConfirmActivity.kt [hunk 15]
```diff
@@ -47,54 +41,74 @@ class PasswordConfirmActivity : SecuredBaseActivity() {
         layout_password_confirm_password.setOnEditorActionListener { _, actionId, _ ->
```
Comment (rmeissner) [inline app/src/main/java/pm/gnosis/heimdall/ui/onboarding/password/PasswordConfirmActivity.kt]: We could use the rxBinding extension and merge the stream with the button clicks stream

**Pull request merged**